MekHQ/src/mekhq/campaign/mission/atb/AtBScenarioFactory.java

```diff
@@ -171,28 +165,20 @@ public static void createScenariosForNewWeek(Campaign c) {
             // Determine active scenarios, to ensure we don't generate a scenario for an already
             // assigned lance and to remove any currently active scenarios from the contract, so that
             // the generation rules are followed for all active scenarios not just new scenarios
-            Iterator<Scenario> iterator = atbContract.getScenarios().iterator();
-            while (iterator.hasNext()) {
-                Scenario scenario = iterator.next();
-                if (!scenario.isCurrent() || !AtBScenario.class.isAssignableFrom(scenario.getClass())) {
-                    continue; //if not current or not assignable to an AtB scenario, then we don't care about it
-                }
-
-                AtBScenario atbScenario = (AtBScenario) scenario;
-
+            for (final AtBScenario scenario : contract.getCurrentAtBScenarios()) {
                 // Add any currently assigned lances to the assignedLances
-                assignedLances.add(atbScenario.getLanceForceId());
+                assignedLances.add(scenario.getLanceForceId());
 
                 // Remove any active scenarios from the contract, and add them to the current scenarios list instead
-                iterator.remove();
-                sList.add(atbScenario);
-                dontGenerateForces.add(atbScenario.getId());
+                contract.getScenarios().remove(scenario);
```

Comment:
Does getCurrentAtBScenarios() return a copy of the collection or something? This seems like it's a great way to get a concurrent modification exception.

Reply:
It does